skfda/misc/scoring.py

```diff
@@ -0,0 +1,1075 @@
+        :class:`~skfda.representation.FDataGrid`
+        objects and multioutput = 'raw_values',
+        :class:`~skfda.representation.FDataGrid` is returned.
+
+        If both :math:`y\_pred` and :math:`y\_true` are ndarray and
+        multioutput = 'raw_values', ndarray.
+
+    """
+    return sklearn.metrics.r2_score(
+        y_true,
+        y_pred,
+        sample_weight=sample_weight,
+        multioutput=multioutput,
+    )
+
+
+@r2_score.register
+def _r2_score_fdatagrid(
+    y_true: FDataGrid,
+    y_pred: FDataGrid,
+    *,
+    sample_weight: Optional[NDArrayFloat] = None,
+    multioutput: Literal['uniform_average', 'raw_values'] = 'uniform_average',
+) -> Union[float, FDataGrid]:
+
+    if y_pred.n_samples < 2:
+        raise ValueError(
+            'R^2 score is not well-defined with less than two samples.',
+        )
+
+    ss_res = mean(
+        np.power(y_true - y_pred, 2),
+        weights=sample_weight,
+    )
+
+    ss_tot = _var(y_true, weights=sample_weight)
+
+    # Divisions by zero allowed
+    with np.errstate(divide='ignore', invalid='ignore'):
+        score = 1 - ss_res / ss_tot
+
+    # 0 / 0 divisions should be 0 in this context and the score, 1
+    score.data_matrix[np.isnan(score.data_matrix)] = 1
+
+    if multioutput == 'raw_values':
+        return score
+
+    # Score only contains 1 function
+    # If the dimension of the codomain is > 1,
+    # the mean of the scores is taken
+    return np.mean(score.integrate()[0]) / _domain_measure(score)
+
+
+@r2_score.register
+def _r2_score_fdatabasis(
+    y_true: FDataBasis,
+    y_pred: FDataBasis,
+    *,
+    sample_weight: Optional[NDArrayFloat] = None,
+) -> float:
+
+    start, end = y_true.domain_range[0]
+
+    if y_pred.n_samples < 2:
+        raise ValueError(
+            'R^2 score is not well-defined with less than two samples.',
+        )
+
+    def _r2_func(x: EvalPointsType) -> NDArrayFloat:  # noqa: WPS430
+        ss_res = np.average(
+            (y_true(x) - y_pred(x)) ** 2,
+            weights=sample_weight,
+            axis=0,
+        )
+
+        ss_tot = np.average(
+            (
+                y_true(x)
+                - np.average(y_true(x), weights=sample_weight, axis=0)
+            ) ** 2,
+            weights=sample_weight,
+            axis=0,
+        )
+
+        # 0/0 case, the score is 1.
+        if ss_res == 0 and ss_tot == 0:
+            return 1
+
+        # r/0 case, r!= 0. Return -inf outside this function
+        if ss_res != 0 and ss_tot == 0:
+            raise ValueError
+
+        score = 1 - ss_res / ss_tot
+
+        # Score only contains 1 function
+        return score[0]
+
+    try:
+        integral = scipy.integrate.quad_vec(
+            _r2_func,
+            start,
+            end,
+        )
+
+    except ValueError:
+        return float('-inf')
+
+    # If the dimension of the codomain is > 1,
+    # the mean of the scores is taken
+    return np.mean(integral[0]) / (end - start)
```

Comment:
Probably this should be a function, as it is common between the different scores. Also, why not `_domain_measure` here?